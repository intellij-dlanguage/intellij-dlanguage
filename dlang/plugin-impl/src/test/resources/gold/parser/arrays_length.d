unittest
{
int[4] foo;
int[]  bar = foo;
int*   p = &foo[0];

// These expressions are equivalent:
bar[];
bar[0 .. 4];
bar[0 .. $];
bar[0 .. bar.length];


p[0 .. $];      // '$' is not defined, since p is not an array
bar[0]+$;            // '$' is not defined, out of scope of [ ]

bar[$-1]; // retrieves last element of the array
}
