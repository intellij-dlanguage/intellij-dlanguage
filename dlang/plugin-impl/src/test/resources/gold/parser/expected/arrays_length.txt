D Language File
  DLanguageDeclarationImpl(DECLARATION)
    UNITTEST
      UNITTEST
        PsiElement(DlangTokenType.unittest)('unittest')
        DLanguageBlockStatementImpl(BLOCK_STATEMENT)
          PsiElement(DlangTokenType.{)('{')
          DLanguageDeclarationsAndStatementsImpl(DECLARATIONS_AND_STATEMENTS)
            DLanguageDeclarationOrStatementImpl(DECLARATION_OR_STATEMENT)
              DLanguageDeclarationImpl(DECLARATION)
                DLanguageVariableDeclarationImpl(VARIABLE_DECLARATION)
                  DLanguageTypeImpl(TYPE)
                    DLanguageType_2Impl(TYPE_2)
                      DLanguageBuiltinTypeImpl(BUILTIN_TYPE)
                        PsiElement(DlangTokenType.int)('int')
                    DLanguageTypeSuffixImpl(TYPE_SUFFIX)
                      PsiElement(DlangTokenType.[)('[')
                      DLanguageAssignExpressionImpl(ASSIGN_EXPRESSION)
                        DLanguageCmpExpressionImpl(CMP_EXPRESSION)
                          DLanguageUnaryExpressionImpl(UNARY_EXPRESSION)
                            DLanguagePrimaryExpressionImpl(PRIMARY_EXPRESSION)
                              PsiElement(DlangTokenType.INTEGER_LITERAL)('4')
                      PsiElement(DlangTokenType.])(']')
                  DECLARATOR
                    IDENTIFIER
                      PsiElement(DlangTokenType.ID)('foo')
                  PsiElement(DlangTokenType.;)(';')
            DLanguageDeclarationOrStatementImpl(DECLARATION_OR_STATEMENT)
              DLanguageDeclarationImpl(DECLARATION)
                DLanguageVariableDeclarationImpl(VARIABLE_DECLARATION)
                  DLanguageTypeImpl(TYPE)
                    DLanguageType_2Impl(TYPE_2)
                      DLanguageBuiltinTypeImpl(BUILTIN_TYPE)
                        PsiElement(DlangTokenType.int)('int')
                    DLanguageTypeSuffixImpl(TYPE_SUFFIX)
                      PsiElement(DlangTokenType.[)('[')
                      PsiElement(DlangTokenType.])(']')
                  DECLARATOR
                    IDENTIFIER
                      PsiElement(DlangTokenType.ID)('bar')
                    PsiElement(DlangTokenType.=)('=')
                    DLanguageInitializerImpl(INITIALIZER)
                      DLanguageNonVoidInitializerImpl(NON_VOID_INITIALIZER)
                        DLanguageAssignExpressionImpl(ASSIGN_EXPRESSION)
                          DLanguageCmpExpressionImpl(CMP_EXPRESSION)
                            DLanguageUnaryExpressionImpl(UNARY_EXPRESSION)
                              DLanguagePrimaryExpressionImpl(PRIMARY_EXPRESSION)
                                DLanguageIdentifierOrTemplateInstanceImpl(IDENTIFIER_OR_TEMPLATE_INSTANCE)
                                  IDENTIFIER
                                    PsiElement(DlangTokenType.ID)('foo')
                  PsiElement(DlangTokenType.;)(';')
            DLanguageDeclarationOrStatementImpl(DECLARATION_OR_STATEMENT)
              DLanguageDeclarationImpl(DECLARATION)
                DLanguageVariableDeclarationImpl(VARIABLE_DECLARATION)
                  DLanguageTypeImpl(TYPE)
                    DLanguageType_2Impl(TYPE_2)
                      DLanguageBuiltinTypeImpl(BUILTIN_TYPE)
                        PsiElement(DlangTokenType.int)('int')
                    DLanguageTypeSuffixImpl(TYPE_SUFFIX)
                      PsiElement(DlangTokenType.*)('*')
                  DECLARATOR
                    IDENTIFIER
                      PsiElement(DlangTokenType.ID)('p')
                    PsiElement(DlangTokenType.=)('=')
                    DLanguageInitializerImpl(INITIALIZER)
                      DLanguageNonVoidInitializerImpl(NON_VOID_INITIALIZER)
                        DLanguageAssignExpressionImpl(ASSIGN_EXPRESSION)
                          DLanguageCmpExpressionImpl(CMP_EXPRESSION)
                            DLanguageUnaryExpressionImpl(UNARY_EXPRESSION)
                              PsiElement(DlangTokenType.&)('&')
                              DLanguageUnaryExpressionImpl(UNARY_EXPRESSION)
                                DLanguagePrimaryExpressionImpl(PRIMARY_EXPRESSION)
                                  DLanguageIdentifierOrTemplateInstanceImpl(IDENTIFIER_OR_TEMPLATE_INSTANCE)
                                    IDENTIFIER
                                      PsiElement(DlangTokenType.ID)('foo')
                                DLanguageIndexExpressionImpl(INDEX_EXPRESSION)
                                  PsiElement(DlangTokenType.[)('[')
                                  DLanguageIndexImpl(INDEX)
                                    DLanguageAssignExpressionImpl(ASSIGN_EXPRESSION)
                                      DLanguageCmpExpressionImpl(CMP_EXPRESSION)
                                        DLanguageUnaryExpressionImpl(UNARY_EXPRESSION)
                                          DLanguagePrimaryExpressionImpl(PRIMARY_EXPRESSION)
                                            PsiElement(DlangTokenType.INTEGER_LITERAL)('0')
                                  PsiElement(DlangTokenType.])(']')
                  PsiElement(DlangTokenType.;)(';')
            PsiComment(DlangTokenType.LINE_COMMENT)('// These expressions are equivalent:')
            DLanguageDeclarationOrStatementImpl(DECLARATION_OR_STATEMENT)
              DLanguageStatementImpl(STATEMENT)
                DLanguageStatementNoCaseNoDefaultImpl(STATEMENT_NO_CASE_NO_DEFAULT)
                  DLanguageExpressionStatementImpl(EXPRESSION_STATEMENT)
                    DLanguageExpressionImpl(EXPRESSION)
                      DLanguageAssignExpressionImpl(ASSIGN_EXPRESSION)
                        DLanguageCmpExpressionImpl(CMP_EXPRESSION)
                          DLanguageUnaryExpressionImpl(UNARY_EXPRESSION)
                            DLanguagePrimaryExpressionImpl(PRIMARY_EXPRESSION)
                              DLanguageIdentifierOrTemplateInstanceImpl(IDENTIFIER_OR_TEMPLATE_INSTANCE)
                                IDENTIFIER
                                  PsiElement(DlangTokenType.ID)('bar')
                            DLanguageIndexExpressionImpl(INDEX_EXPRESSION)
                              PsiElement(DlangTokenType.[)('[')
                              PsiElement(DlangTokenType.])(']')
                    PsiElement(DlangTokenType.;)(';')
            DLanguageDeclarationOrStatementImpl(DECLARATION_OR_STATEMENT)
              DLanguageStatementImpl(STATEMENT)
                DLanguageStatementNoCaseNoDefaultImpl(STATEMENT_NO_CASE_NO_DEFAULT)
                  DLanguageExpressionStatementImpl(EXPRESSION_STATEMENT)
                    DLanguageExpressionImpl(EXPRESSION)
                      DLanguageAssignExpressionImpl(ASSIGN_EXPRESSION)
                        DLanguageCmpExpressionImpl(CMP_EXPRESSION)
                          DLanguageUnaryExpressionImpl(UNARY_EXPRESSION)
                            DLanguagePrimaryExpressionImpl(PRIMARY_EXPRESSION)
                              DLanguageIdentifierOrTemplateInstanceImpl(IDENTIFIER_OR_TEMPLATE_INSTANCE)
                                IDENTIFIER
                                  PsiElement(DlangTokenType.ID)('bar')
                            DLanguageIndexExpressionImpl(INDEX_EXPRESSION)
                              PsiElement(DlangTokenType.[)('[')
                              DLanguageIndexImpl(INDEX)
                                DLanguageAssignExpressionImpl(ASSIGN_EXPRESSION)
                                  DLanguageCmpExpressionImpl(CMP_EXPRESSION)
                                    DLanguageUnaryExpressionImpl(UNARY_EXPRESSION)
                                      DLanguagePrimaryExpressionImpl(PRIMARY_EXPRESSION)
                                        PsiElement(DlangTokenType.INTEGER_LITERAL)('0')
                                PsiElement(DlangTokenType...)('..')
                                DLanguageAssignExpressionImpl(ASSIGN_EXPRESSION)
                                  DLanguageCmpExpressionImpl(CMP_EXPRESSION)
                                    DLanguageUnaryExpressionImpl(UNARY_EXPRESSION)
                                      DLanguagePrimaryExpressionImpl(PRIMARY_EXPRESSION)
                                        PsiElement(DlangTokenType.INTEGER_LITERAL)('4')
                              PsiElement(DlangTokenType.])(']')
                    PsiElement(DlangTokenType.;)(';')
            DLanguageDeclarationOrStatementImpl(DECLARATION_OR_STATEMENT)
              DLanguageStatementImpl(STATEMENT)
                DLanguageStatementNoCaseNoDefaultImpl(STATEMENT_NO_CASE_NO_DEFAULT)
                  DLanguageExpressionStatementImpl(EXPRESSION_STATEMENT)
                    DLanguageExpressionImpl(EXPRESSION)
                      DLanguageAssignExpressionImpl(ASSIGN_EXPRESSION)
                        DLanguageCmpExpressionImpl(CMP_EXPRESSION)
                          DLanguageUnaryExpressionImpl(UNARY_EXPRESSION)
                            DLanguagePrimaryExpressionImpl(PRIMARY_EXPRESSION)
                              DLanguageIdentifierOrTemplateInstanceImpl(IDENTIFIER_OR_TEMPLATE_INSTANCE)
                                IDENTIFIER
                                  PsiElement(DlangTokenType.ID)('bar')
                            DLanguageIndexExpressionImpl(INDEX_EXPRESSION)
                              PsiElement(DlangTokenType.[)('[')
                              DLanguageIndexImpl(INDEX)
                                DLanguageAssignExpressionImpl(ASSIGN_EXPRESSION)
                                  DLanguageCmpExpressionImpl(CMP_EXPRESSION)
                                    DLanguageUnaryExpressionImpl(UNARY_EXPRESSION)
                                      DLanguagePrimaryExpressionImpl(PRIMARY_EXPRESSION)
                                        PsiElement(DlangTokenType.INTEGER_LITERAL)('0')
                                PsiElement(DlangTokenType...)('..')
                                DLanguageAssignExpressionImpl(ASSIGN_EXPRESSION)
                                  DLanguageCmpExpressionImpl(CMP_EXPRESSION)
                                    DLanguageUnaryExpressionImpl(UNARY_EXPRESSION)
                                      DLanguagePrimaryExpressionImpl(PRIMARY_EXPRESSION)
                                        PsiElement(DlangTokenType.$)('$')
                              PsiElement(DlangTokenType.])(']')
                    PsiElement(DlangTokenType.;)(';')
            DLanguageDeclarationOrStatementImpl(DECLARATION_OR_STATEMENT)
              DLanguageStatementImpl(STATEMENT)
                DLanguageStatementNoCaseNoDefaultImpl(STATEMENT_NO_CASE_NO_DEFAULT)
                  DLanguageExpressionStatementImpl(EXPRESSION_STATEMENT)
                    DLanguageExpressionImpl(EXPRESSION)
                      DLanguageAssignExpressionImpl(ASSIGN_EXPRESSION)
                        DLanguageCmpExpressionImpl(CMP_EXPRESSION)
                          DLanguageUnaryExpressionImpl(UNARY_EXPRESSION)
                            DLanguagePrimaryExpressionImpl(PRIMARY_EXPRESSION)
                              DLanguageIdentifierOrTemplateInstanceImpl(IDENTIFIER_OR_TEMPLATE_INSTANCE)
                                IDENTIFIER
                                  PsiElement(DlangTokenType.ID)('bar')
                            DLanguageIndexExpressionImpl(INDEX_EXPRESSION)
                              PsiElement(DlangTokenType.[)('[')
                              DLanguageIndexImpl(INDEX)
                                DLanguageAssignExpressionImpl(ASSIGN_EXPRESSION)
                                  DLanguageCmpExpressionImpl(CMP_EXPRESSION)
                                    DLanguageUnaryExpressionImpl(UNARY_EXPRESSION)
                                      DLanguagePrimaryExpressionImpl(PRIMARY_EXPRESSION)
                                        PsiElement(DlangTokenType.INTEGER_LITERAL)('0')
                                PsiElement(DlangTokenType...)('..')
                                DLanguageAssignExpressionImpl(ASSIGN_EXPRESSION)
                                  DLanguageCmpExpressionImpl(CMP_EXPRESSION)
                                    DLanguageUnaryExpressionImpl(UNARY_EXPRESSION)
                                      DLanguagePrimaryExpressionImpl(PRIMARY_EXPRESSION)
                                        DLanguageIdentifierOrTemplateInstanceImpl(IDENTIFIER_OR_TEMPLATE_INSTANCE)
                                          IDENTIFIER
                                            PsiElement(DlangTokenType.ID)('bar')
                                      PsiElement(DlangTokenType..)('.')
                                      DLanguageIdentifierOrTemplateInstanceImpl(IDENTIFIER_OR_TEMPLATE_INSTANCE)
                                        IDENTIFIER
                                          PsiElement(DlangTokenType.ID)('length')
                              PsiElement(DlangTokenType.])(']')
                    PsiElement(DlangTokenType.;)(';')
            DLanguageDeclarationOrStatementImpl(DECLARATION_OR_STATEMENT)
              DLanguageStatementImpl(STATEMENT)
                DLanguageStatementNoCaseNoDefaultImpl(STATEMENT_NO_CASE_NO_DEFAULT)
                  DLanguageExpressionStatementImpl(EXPRESSION_STATEMENT)
                    DLanguageExpressionImpl(EXPRESSION)
                      DLanguageAssignExpressionImpl(ASSIGN_EXPRESSION)
                        DLanguageCmpExpressionImpl(CMP_EXPRESSION)
                          DLanguageUnaryExpressionImpl(UNARY_EXPRESSION)
                            DLanguagePrimaryExpressionImpl(PRIMARY_EXPRESSION)
                              DLanguageIdentifierOrTemplateInstanceImpl(IDENTIFIER_OR_TEMPLATE_INSTANCE)
                                IDENTIFIER
                                  PsiElement(DlangTokenType.ID)('p')
                            DLanguageIndexExpressionImpl(INDEX_EXPRESSION)
                              PsiElement(DlangTokenType.[)('[')
                              DLanguageIndexImpl(INDEX)
                                DLanguageAssignExpressionImpl(ASSIGN_EXPRESSION)
                                  DLanguageCmpExpressionImpl(CMP_EXPRESSION)
                                    DLanguageUnaryExpressionImpl(UNARY_EXPRESSION)
                                      DLanguagePrimaryExpressionImpl(PRIMARY_EXPRESSION)
                                        PsiElement(DlangTokenType.INTEGER_LITERAL)('0')
                                PsiElement(DlangTokenType...)('..')
                                DLanguageAssignExpressionImpl(ASSIGN_EXPRESSION)
                                  DLanguageCmpExpressionImpl(CMP_EXPRESSION)
                                    DLanguageUnaryExpressionImpl(UNARY_EXPRESSION)
                                      DLanguagePrimaryExpressionImpl(PRIMARY_EXPRESSION)
                                        PsiElement(DlangTokenType.$)('$')
                              PsiElement(DlangTokenType.])(']')
                    PsiElement(DlangTokenType.;)(';')
            PsiComment(DlangTokenType.LINE_COMMENT)('// '$' is not defined, since p is not an array')
            DLanguageDeclarationOrStatementImpl(DECLARATION_OR_STATEMENT)
              DLanguageStatementImpl(STATEMENT)
                DLanguageStatementNoCaseNoDefaultImpl(STATEMENT_NO_CASE_NO_DEFAULT)
                  DLanguageExpressionStatementImpl(EXPRESSION_STATEMENT)
                    DLanguageExpressionImpl(EXPRESSION)
                      DLanguageAssignExpressionImpl(ASSIGN_EXPRESSION)
                        DLanguageCmpExpressionImpl(CMP_EXPRESSION)
                          DLanguageAddExpressionImpl(ADD_EXPRESSION)
                            DLanguageUnaryExpressionImpl(UNARY_EXPRESSION)
                              DLanguagePrimaryExpressionImpl(PRIMARY_EXPRESSION)
                                DLanguageIdentifierOrTemplateInstanceImpl(IDENTIFIER_OR_TEMPLATE_INSTANCE)
                                  IDENTIFIER
                                    PsiElement(DlangTokenType.ID)('bar')
                              DLanguageIndexExpressionImpl(INDEX_EXPRESSION)
                                PsiElement(DlangTokenType.[)('[')
                                DLanguageIndexImpl(INDEX)
                                  DLanguageAssignExpressionImpl(ASSIGN_EXPRESSION)
                                    DLanguageCmpExpressionImpl(CMP_EXPRESSION)
                                      DLanguageUnaryExpressionImpl(UNARY_EXPRESSION)
                                        DLanguagePrimaryExpressionImpl(PRIMARY_EXPRESSION)
                                          PsiElement(DlangTokenType.INTEGER_LITERAL)('0')
                                PsiElement(DlangTokenType.])(']')
                            PsiElement(DlangTokenType.+)('+')
                            DLanguageUnaryExpressionImpl(UNARY_EXPRESSION)
                              DLanguagePrimaryExpressionImpl(PRIMARY_EXPRESSION)
                                PsiElement(DlangTokenType.$)('$')
                    PsiElement(DlangTokenType.;)(';')
            PsiComment(DlangTokenType.LINE_COMMENT)('// '$' is not defined, out of scope of [ ]')
            DLanguageDeclarationOrStatementImpl(DECLARATION_OR_STATEMENT)
              DLanguageStatementImpl(STATEMENT)
                DLanguageStatementNoCaseNoDefaultImpl(STATEMENT_NO_CASE_NO_DEFAULT)
                  DLanguageExpressionStatementImpl(EXPRESSION_STATEMENT)
                    DLanguageExpressionImpl(EXPRESSION)
                      DLanguageAssignExpressionImpl(ASSIGN_EXPRESSION)
                        DLanguageCmpExpressionImpl(CMP_EXPRESSION)
                          DLanguageUnaryExpressionImpl(UNARY_EXPRESSION)
                            DLanguagePrimaryExpressionImpl(PRIMARY_EXPRESSION)
                              DLanguageIdentifierOrTemplateInstanceImpl(IDENTIFIER_OR_TEMPLATE_INSTANCE)
                                IDENTIFIER
                                  PsiElement(DlangTokenType.ID)('bar')
                            DLanguageIndexExpressionImpl(INDEX_EXPRESSION)
                              PsiElement(DlangTokenType.[)('[')
                              DLanguageIndexImpl(INDEX)
                                DLanguageAssignExpressionImpl(ASSIGN_EXPRESSION)
                                  DLanguageCmpExpressionImpl(CMP_EXPRESSION)
                                    DLanguageAddExpressionImpl(ADD_EXPRESSION)
                                      DLanguageUnaryExpressionImpl(UNARY_EXPRESSION)
                                        DLanguagePrimaryExpressionImpl(PRIMARY_EXPRESSION)
                                          PsiElement(DlangTokenType.$)('$')
                                      PsiElement(DlangTokenType.-)('-')
                                      DLanguageUnaryExpressionImpl(UNARY_EXPRESSION)
                                        DLanguagePrimaryExpressionImpl(PRIMARY_EXPRESSION)
                                          PsiElement(DlangTokenType.INTEGER_LITERAL)('1')
                              PsiElement(DlangTokenType.])(']')
                    PsiElement(DlangTokenType.;)(';')
          PsiComment(DlangTokenType.LINE_COMMENT)('// retrieves last element of the array')
          PsiElement(DlangTokenType.})('}')